subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultIvyArtifactRepository.java

```diff
@@ -221,9 +221,13 @@ private void validate(Set<String> schemes) {
     private Set<String> getSchemes() {
         if (schemes == null) {
             URI uri = getUrl();
-            schemes = new LinkedHashSet<>();
-            layout.addSchemes(uri, schemes);
-            additionalPatternsLayout.addSchemes(uri, schemes);
+            // use a local variable to prepare the set,
+            // so that other threads do not see the half-initialized
+            // list of schemes and fail in strange ways
```

Comment:
This is not sufficient to prevent race conditions, as the variable itself is not `volatile`. I think this should be fine if `schemes` is changed to be `volatile`, as it seems the initialization is pure so running it more than once will produce the same result. If this is not the case, it will need full double-checked locking.

Reply:
Adding `volatile` would make the threads sync on the value before accessing it.
Like it is currently, if a thread does not see the value another thread already set it will recalculate the value.
Which should not be a problem as the value should be reproducible and not change.
So different threads might use different identical lists but that should be ok, shouldn't it?

Adding `volatile` would reduce the amount of times this initialization is done for the price that the field is synchronized on each read.

The problem with the previous implementation was, that if thread A set the empty list, then thread B is started and thus has that value for the field and sees the empty list.

Don't you think it should be sufficient like it is?

If not, we should probably also use an immutable list as value just to make sure no thread is modifying it as the current instance also is not thread-safe if it would be modified by some code.

Reply:
Yes, `volatile` would make the threads sync, though in practice on x86 this isn't very noticeable and on ARM it should be pretty quick still, I don't think this is a highly contested field.

It is necessary to sync because the fields of an object are subject to different memory synchronization than the reference itself, i.e. without `volatile` it is still perfectly legal for thread A to assign the field to the object reference, and finish initializing it _after_ B has read the object reference as per https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/#wishful-benign-is-resilient.

I am all for using an immutable instance, though depending on exactly how it is constructed it may still require `volatile` to ensure thread safety.

Reply:
Ok, I added `volatile` and the immutability now. :-)